udmis/src/main/java/com/google/bos/udmi/monitoring/LogTail.java

```diff
@@ -0,0 +1,253 @@
+    LoggingOptions options = LoggingOptions.getDefaultInstance();
+    Logging logging = options.getService();
+    LogEntryServerStream stream;
+    if (filter == null) {
+      stream = logging.tailLogEntries();
+    } else {
+      stream = logging.tailLogEntries(TailOption.filter(filter));
+    }
+    return stream;
+  }
+
+  /*
+   * Top level processor of incoming GCP log entries.
+   */
+  private void processLogEntry(LogEntry log) {
+    if (log.getSeverity().equals(Severity.ERROR)) {
+      processLogEntryError(log);
+    } else {
+      String function_name = log.getResource().getLabels().getOrDefault("function_name", null);
+      if (function_name.equals("udmi_target")) {
+      } else if (function_name.equals("udmi_state")) {
+      } else if (function_name.equals("udmi_config")) {
+      }
+    }
+  }
+
+  /*
+   * Top level processor of incoming GCP log entries of ERROR type.
+   */
+  private void processLogEntryError(LogEntry log) {
+    // Be careful in here. Unwrapping this is a careful sequence.
+    Payload.ProtoPayload payload = log.getPayload();
+    if (!payload.getData().getTypeUrl().equals("type.googleapis.com/google.cloud.audit.AuditLog")) {
+      System.err.println("Not AuditLog");
+      return;
+    }
+    try {
+      AuditLog auditLog = AuditLog.parseFrom(payload.getData().getValue().toByteArray());
+      LogTailEntry entry = new LogTailEntry();
+      entry.timestamp = log.getInstantTimestamp();
+      entry.methodName = auditLog.getMethodName();
+      entry.serviceName = auditLog.getServiceName();
+      entry.resourceName = auditLog.getResourceName();
+      entry.statusCode = auditLog.getStatus().getCode();
+      entry.statusMessage = auditLog.getStatus().getMessage();
+      entry.severity = log.getSeverity();
+      logsTimeSeries.add(entry);
+      logsTimeSeries.maybeEmitMetrics();
+    } catch (InvalidProtocolBufferException e) {
+      System.err.println(e.toString());
+    } catch (IOException e) {
+      System.err.println(e.toString());
+    }
+  }
+
+  private void processLogStream(LogEntryServerStream stream) {
+    for (LogEntry log : stream) {
+      processLogEntry(log);
+    }
+  }
+
+  public void tailLogs() {
+    LogEntryServerStream stream = getCloudLogStream(LOG_FILTER);
+    processLogStream(stream);
+  }
+
+  /*
+   * Fields from GCP Logging stored in a neutral class.
+   */
+  protected class LogTailEntry {
+
+    Instant timestamp;
+    String methodName;
+    String serviceName;
+    String resourceName;
+    int statusCode;
+    String statusMessage;
+    Severity severity;
+    int value = 1;
+  }
+
+  /*
+   * Linked list of LogTailEntry.
+   */
+  protected class LogTailEntryList extends LinkedList<LogTailEntry> {
+
+  }
+
+  /*
+   * All the LogTail items currently processing.
+   * Hold the log entries in buckets for upstream processing in chunks.
+   */
+  protected class LogTimeSeries extends TreeMap<Long, LogTailEntryList> {
```

Comment:
This seems like a lot for an inner class.  Can it be made standalone?  Otherwise can it be static?  What from the containing class is it referencing?